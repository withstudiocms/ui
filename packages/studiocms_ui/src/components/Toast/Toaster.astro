---
import './toaster.css';

/**
 * Props for the Toast component.
 */
interface Props {
	/**
	 * The position of the toaster. All toasts will originate from this position.
	 */
	position?:
		| 'top-left'
		| 'top-right'
		| 'top-center'
		| 'bottom-left'
		| 'bottom-right'
		| 'bottom-center';
	/**
	 * The duration of the toast in milliseconds. Defaults to 4000 (4 seconds).
	 */
	duration?: number;
	/**
	 * Whether the toast has a close button. Defaults to false.
	 */
	closeButton?: boolean;
	/**
	 * The offset of the toaster from the edge of the screen in pixels. Defaults to 32.
	 */
	offset?: number;
	/**
	 * The gap between toasts in pixels. Defaults to 8.
	 */
	gap?: number;
}

const {
	position = 'top-center',
	duration = 4000,
	closeButton = false,
	offset = 32,
	gap = 8,
} = Astro.props;
---
<div 
  id="sui-toaster" 
  class:list={[
    closeButton && "closeable",
    position,
  ]},
>
  <div 
    id="sui-toast-drawer"
    data-offset={offset}
    data-gap={gap}
    data-duration={duration}
    style={[
      `${position.includes("top-") ? 'top:' : 'bottom:'} ${offset}px;`,
      position.includes("-left") && `left: ${offset}px`,
      position.includes("-right") && `right: ${offset}px`,
      position.includes("-center") && `left: 50%; transform: translateX(-50%);`,
      `--gap: ${gap}px;`,
      `padding-left: ${offset}px;`,
      `padding-right: ${offset}px;`,
    ].filter(Boolean).join("")}
  />
</div>
<script>
import "studiocms:ui/scripts/toaster"
</script>
