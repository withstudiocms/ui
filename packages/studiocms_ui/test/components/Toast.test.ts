import { describe, expect } from 'vitest';
import Toaster from '../../src/components/Toast/Toaster.astro';
import { test } from '../fixtures/vitest/AstroContainer';

describe('Toaster Component', () => {
	test('renders Toaster component correctly', async ({ renderComponent }) => {
		const result = await renderComponent(Toaster, 'Toaster');
		expect(result).toMatchSnapshot();
	});

	test.for([
		{ position: 'top-left' },
		{ position: 'top-center' },
		{ position: 'top-right' },
		{ position: 'bottom-left' },
		{ position: 'bottom-center' },
		{ position: 'bottom-right' },
		{ duration: 5000 },
		{ duration: 10000 },
		{ closeButton: true },
		{ closeButton: false },
		{ offset: 20 },
		{ gap: 15 },
	])('renders Toaster component with props "%s"', async (props, { renderComponent }) => {
		const result = await renderComponent(Toaster, 'Toaster', { props });
		expect(result).toMatchSnapshot();
	});
});
