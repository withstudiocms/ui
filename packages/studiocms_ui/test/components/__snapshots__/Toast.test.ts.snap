// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toaster Component > renders Toaster component correctly 1`] = `"<div id="sui-toaster" class="top-center"> <div id="sui-toast-drawer" data-offset="32" data-gap="8" data-duration="4000" style="top: 32px;left: 50%; transform: translateX(-50%);--gap: 8px;padding-left: 32px;padding-right: 32px;"></div> </div> <script type="module" src="/mock/path/Toaster.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toaster Component > renders Toaster component with props "{ closeButton: false }" 1`] = `"<div id="sui-toaster" class="top-center"> <div id="sui-toast-drawer" data-offset="32" data-gap="8" data-duration="4000" style="top: 32px;left: 50%; transform: translateX(-50%);--gap: 8px;padding-left: 32px;padding-right: 32px;"></div> </div> <script type="module" src="/mock/path/Toaster.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toaster Component > renders Toaster component with props "{ closeButton: true }" 1`] = `"<div id="sui-toaster" class="closeable top-center"> <div id="sui-toast-drawer" data-offset="32" data-gap="8" data-duration="4000" style="top: 32px;left: 50%; transform: translateX(-50%);--gap: 8px;padding-left: 32px;padding-right: 32px;"></div> </div> <script type="module" src="/mock/path/Toaster.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toaster Component > renders Toaster component with props "{ duration: 5000 }" 1`] = `"<div id="sui-toaster" class="top-center"> <div id="sui-toast-drawer" data-offset="32" data-gap="8" data-duration="5000" style="top: 32px;left: 50%; transform: translateX(-50%);--gap: 8px;padding-left: 32px;padding-right: 32px;"></div> </div> <script type="module" src="/mock/path/Toaster.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toaster Component > renders Toaster component with props "{ duration: 10000 }" 1`] = `"<div id="sui-toaster" class="top-center"> <div id="sui-toast-drawer" data-offset="32" data-gap="8" data-duration="10000" style="top: 32px;left: 50%; transform: translateX(-50%);--gap: 8px;padding-left: 32px;padding-right: 32px;"></div> </div> <script type="module" src="/mock/path/Toaster.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toaster Component > renders Toaster component with props "{ gap: 15 }" 1`] = `"<div id="sui-toaster" class="top-center"> <div id="sui-toast-drawer" data-offset="32" data-gap="15" data-duration="4000" style="top: 32px;left: 50%; transform: translateX(-50%);--gap: 15px;padding-left: 32px;padding-right: 32px;"></div> </div> <script type="module" src="/mock/path/Toaster.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toaster Component > renders Toaster component with props "{ offset: 20 }" 1`] = `"<div id="sui-toaster" class="top-center"> <div id="sui-toast-drawer" data-offset="20" data-gap="8" data-duration="4000" style="top: 20px;left: 50%; transform: translateX(-50%);--gap: 8px;padding-left: 20px;padding-right: 20px;"></div> </div> <script type="module" src="/mock/path/Toaster.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toaster Component > renders Toaster component with props "{ position: 'bottom-center' }" 1`] = `"<div id="sui-toaster" class="bottom-center"> <div id="sui-toast-drawer" data-offset="32" data-gap="8" data-duration="4000" style="bottom: 32px;left: 50%; transform: translateX(-50%);--gap: 8px;padding-left: 32px;padding-right: 32px;"></div> </div> <script type="module" src="/mock/path/Toaster.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toaster Component > renders Toaster component with props "{ position: 'bottom-left' }" 1`] = `"<div id="sui-toaster" class="bottom-left"> <div id="sui-toast-drawer" data-offset="32" data-gap="8" data-duration="4000" style="bottom: 32px;left: 32px--gap: 8px;padding-left: 32px;padding-right: 32px;"></div> </div> <script type="module" src="/mock/path/Toaster.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toaster Component > renders Toaster component with props "{ position: 'bottom-right' }" 1`] = `"<div id="sui-toaster" class="bottom-right"> <div id="sui-toast-drawer" data-offset="32" data-gap="8" data-duration="4000" style="bottom: 32px;right: 32px--gap: 8px;padding-left: 32px;padding-right: 32px;"></div> </div> <script type="module" src="/mock/path/Toaster.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toaster Component > renders Toaster component with props "{ position: 'top-center' }" 1`] = `"<div id="sui-toaster" class="top-center"> <div id="sui-toast-drawer" data-offset="32" data-gap="8" data-duration="4000" style="top: 32px;left: 50%; transform: translateX(-50%);--gap: 8px;padding-left: 32px;padding-right: 32px;"></div> </div> <script type="module" src="/mock/path/Toaster.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toaster Component > renders Toaster component with props "{ position: 'top-left' }" 1`] = `"<div id="sui-toaster" class="top-left"> <div id="sui-toast-drawer" data-offset="32" data-gap="8" data-duration="4000" style="top: 32px;left: 32px--gap: 8px;padding-left: 32px;padding-right: 32px;"></div> </div> <script type="module" src="/mock/path/Toaster.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toaster Component > renders Toaster component with props "{ position: 'top-right' }" 1`] = `"<div id="sui-toaster" class="top-right"> <div id="sui-toast-drawer" data-offset="32" data-gap="8" data-duration="4000" style="top: 32px;right: 32px--gap: 8px;padding-left: 32px;padding-right: 32px;"></div> </div> <script type="module" src="/mock/path/Toaster.astro?astro&type=script&index=0&lang.ts"></script>"`;
